sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala

```diff
@@ -195,9 +199,7 @@ trait ColumnResolutionHelper extends Logging {
     }
 
     try {
-      val resolved = innerResolve(expr, isTopLevel = true)
-      val withOuterResolved = if (allowOuter) resolveOuterRef(resolved) else resolved
-      resolveVariables(withOuterResolved)
```

Comment:
This is the bug. Sometimes we only want to do basic column resolution first, and then try some special resolution logic (see `ResolveReferencesInAggregat` as an example), and finally resolve outer references. The code before always tried to resolve the SQL variable.

To avoid hitting similar bugs, I refactored the code a bit to have a `last resort` step for column resolution, which contains resolving outer references and SQL variables.